be/src/exec/hash_join_components.cpp

```diff
@@ -155,4 +400,387 @@ ChunkPtr SingleHashJoinBuilder::convert_to_spill_schema(const ChunkPtr& chunk) c
+    Status _convert_to_single_partition();
+    Status _append_chunk_to_partitions(const ChunkPtr& chunk);
+
+private:
+    std::vector<std::unique_ptr<SingleHashJoinBuilder>> _builders;
+
+    size_t _partition_num = 0;
+    size_t _partition_join_min_rows = 0;
+    size_t _partition_join_max_rows = 0;
+
+    size_t _probe_estimated_costs = 0;
+
+    size_t _fit_L2_cache_max_rows = 0;
+    size_t _fit_L3_cache_max_rows = 0;
+
+    size_t _L2_cache_size = 0;
+    size_t _L3_cache_size = 0;
+
+    size_t _pushed_chunks = 0;
+};
+
+AdaptivePartitionHashJoinBuilder::AdaptivePartitionHashJoinBuilder(HashJoiner& hash_joiner)
+        : HashJoinBuilder(hash_joiner) {
+    const size_t DEFAULT_L2_CACHE_SIZE = 1 * 1024 * 1024;
+    const size_t DEFAULT_L3_CACHE_SIZE = 32 * 1024 * 1024;
+    const auto& cache_sizes = CpuInfo::get_cache_sizes();
+    _L2_cache_size = cache_sizes[CpuInfo::L2_CACHE];
+    _L3_cache_size = cache_sizes[CpuInfo::L3_CACHE];
+    _L2_cache_size = _L2_cache_size ? _L2_cache_size : DEFAULT_L2_CACHE_SIZE;
+    _L3_cache_size = _L3_cache_size ? _L3_cache_size : DEFAULT_L3_CACHE_SIZE;
+}
+
+size_t AdaptivePartitionHashJoinBuilder::_estimated_row_size(const HashTableParam& param) const {
+    size_t estimated_each_row = 0;
+
+    for (auto* tuple : param.build_row_desc->tuple_descriptors()) {
+        for (auto slot : tuple->slots()) {
+            if (param.build_output_slots.contains(slot->id())) {
+                estimated_each_row += get_size_of_fixed_length_type(slot->type().type);
+                estimated_each_row += type_estimated_overhead_bytes(slot->type().type);
+            }
+        }
+    }
+
+    // for hash table bucket
+    estimated_each_row += 4;
+
+    return estimated_each_row;
+}
+
+// We could use a better estimation model.
+size_t AdaptivePartitionHashJoinBuilder::_estimated_probe_cost(const HashTableParam& param) const {
+    size_t size = 0;
+
+    for (auto* tuple : param.probe_row_desc->tuple_descriptors()) {
+        for (auto slot : tuple->slots()) {
+            if (param.probe_output_slots.contains(slot->id())) {
+                size += get_size_of_fixed_length_type(slot->type().type);
+                size += type_estimated_overhead_bytes(slot->type().type);
+            }
+        }
+    }
+    // we define probe cost is bytes size * 6
+    return size * 6;
+}
+
+template <>
+size_t AdaptivePartitionHashJoinBuilder::_estimated_build_cost<CacheLevel::L2>(size_t build_row_size) const {
+    return build_row_size / 2;
+}
+
+template <>
+size_t AdaptivePartitionHashJoinBuilder::_estimated_build_cost<CacheLevel::L3>(size_t build_row_size) const {
+    return build_row_size;
+}
+
+template <>
+size_t AdaptivePartitionHashJoinBuilder::_estimated_build_cost<CacheLevel::MEMORY>(size_t build_row_size) const {
+    return build_row_size * 2;
+}
+
+void AdaptivePartitionHashJoinBuilder::_adjust_partition_rows(size_t build_row_size) {
+    _fit_L2_cache_max_rows = _L2_cache_size / build_row_size;
```

Comment:
Maybe =0 ?

Reply:
build_row_size will never be zero
```
    size_t estimated_each_row = 0;
    ...

    // for hash table bucket
    estimated_each_row += 4;
```

but we still could add a protection code